File: main/efficiently-sorting-number-from-0-to-n2-1/efficiently-sorting-number-from-0-to-n2-1.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.StringJoiner;
import java.util.function.IntUnaryOperator;
import java.util.stream.IntStream;

final class Solution {
    // Produces an array (of length size) of values, scanned from sc.
    private static int[] readRecord(final Scanner sc, final int size)
    {
        return IntStream.range(0, size).map(i -> sc.nextInt()).toArray();
    }
    
    // Prints all the numbers in an array.
    private static void print(final int[] values)
    {
        final StringJoiner sj = new StringJoiner(" ");
        for (final int x : values) sj.add(String.valueOf(x));
        System.out.println(sj);
    }
    
    // Makes a list of nulls representing empty bins. [countingSort() helper.]
    private static List<List<Integer>> initializeEmptyBins(int keyRange)
    {
        if (keyRange < 0) throw new IllegalArgumentException();
        
        final List<List<Integer>> bins = new ArrayList<>(keyRange);
        while (keyRange-- != 0) bins.add(null);
        return bins;
    }
    
    // Get a bin, creating it if necessary. [countingSort() helper.]
    private static List<Integer> getBin(final List<List<Integer>> bins,
                                        final int key)
    {
        List<Integer> ret = bins.get(key);
        
        if (ret == null) {
            ret = new ArrayList<>();
            bins.set(key, ret);
        }
        
        return ret;
    }
    
    // Stably group items into bins by key. [countingSort() helper.]
    private static List<List<Integer>> group(final int[] values,
            final int keyRange, final IntUnaryOperator keySelector)
    {
        final List<List<Integer>> bins = initializeEmptyBins(keyRange);
        
        for (final int x : values)
            getBin(bins, keySelector.applyAsInt(x)).add(x);
        
        return bins;
    }
    
    // Do a counting sort, selecting keys known to be in a prescribed range.
    private static void countingSort(final int[] values, // mutates values
            final int keyRange, final IntUnaryOperator keySelector)
    {
        int i = 0;
        for (final List<Integer> bin : group(values, keyRange, keySelector)) {
            if (bin != null)
                for (final int x : bin) values[i++] = x;
        }
        
        assert i == values.length;
    }
    
    // Do a two-pass radix sort, using countingSort().
    // Assumes every element of values is strictly less than radix * radix.
    private static void radixSort(final int[] values, final int radix)
    {
        if (radix <= 0) throw new IllegalArgumentException();
        
        countingSort(values, radix, x -> x % radix);
        countingSort(values, radix, x -> x / radix);
    }
    
    public static void main(final String[] args)
    {
        try (final Scanner sc = new Scanner(System.in)) {
            for (int t = sc.nextInt(); t > 0; --t) {
                final int size = sc.nextInt();
                final int[] values = readRecord(sc, size);
                radixSort(values, size * size);
                print(values);
            }
        }
    }
    
    private Solution() { } // the Solution class should not be instantiated
}

File: main/efficiently-sorting-number-from-0-to-n2-1/efficiently-sorting-number-from-0-to-n2-1.cpp
#include <cassert>
#include <cstddef>
#include <vector>
#include <iostream>
#include <iterator>
#include <stdexcept>

namespace {
    int check_size_and_get_square(const int size) // helper for read_record()
    {
        if (size < 0) throw std::out_of_range{"size must be nonnegative"};

        const auto square = size * size;

        if (square < 0 || square / size != size)
            throw std::out_of_range{"size too big"};

        return square;
    }

    std::vector<int> read_record(const int size)
    {
        const auto square = check_size_and_get_square(size);
        
        std::vector<int> values (static_cast<std::size_t>(size));
        
        for (auto& x : values) {
            std::cin >> x;
            if (x >= square) throw std::out_of_range{"element too big"};
        }
        
        return values;
    }
    
    void print(const std::vector<int>& values)
    {
        auto sep = "";
        for (const auto x : values) {
            std::cout << sep << x;
            sep = " ";
        }
        
        std::cout << '\n';
    }
    
    template<class TFunc> // helper function for counting_sort()
    std::vector<std::vector<int>> counting_sort_bins(
            const std::vector<int>& values, const std::size_t key_range,
                                            const TFunc key_selector)
    {
        std::vector<std::vector<int>> bins (key_range);
        for (const auto x : values) bins[key_selector(x)].push_back(x);
        return bins;
    }
    
    template<class TFunc>
    void counting_sort(std::vector<int>& values, const std::size_t key_range,
                       const TFunc key_selector)
    {
        const auto bins = counting_sort_bins(values, key_range, key_selector);
        auto p = std::begin(values);
        
        for (const auto& bin : bins) {
            for (const auto x : bin) {
                *p = x;
                ++p;
            }
        }
        
        assert(p == std::end(values));
    }
    
    // assumes every element of values is strictly less than radix * radix
    void radix_sort(std::vector<int>& values, const int radix)
    {
        assert(radix > 0);
        const auto key_range = static_cast<std::size_t>(radix);
        
        counting_sort(values, key_range, [radix](const int x) {
            return static_cast<std::size_t>(x % radix);
        });
        
        counting_sort(values, key_range, [radix](const int x) {
            return static_cast<std::size_t>(x / radix);
        });
    }
}

int main()
{
    std::ios_base::sync_with_stdio(false);
    
    auto t = 0;
    for (std::cin >> t; t > 0; --t) {
        auto size = 0;
        std::cin >> size;
        
        auto values = read_record(size);
        radix_sort(values, size * size);
        print(values);
    }
}
